File: Sampler.jl
using PyPlot

@doc """
Sampler class
""" ->
type Sampler
	num_particles::Int64
	mcmc_steps::Int64
	particles::Vector{Particle}
	logl::Vector{Float64}
	tiebreakers::Vector{Float64}

	# Current iteration
	iteration::Int64

	# Current log likelihood threshold
	logx_threshold::Float64
	logl_threshold::Float64
	tb_threshold::Float64
end

@doc """
Constructor that only takes num_particles and mcmc_steps
as input
""" ->
function Sampler(num_particles::Int64, mcmc_steps::Int64)
	@assert (num_particles >= 1) & (mcmc_steps >= 1)
	return Sampler(num_particles, mcmc_steps,
								Array(Particle, (num_particles, )),
								zeros(num_particles), zeros(num_particles),
								0, 0.0, -Inf, 0.0)
end

@doc """
Generate all particles from the prior
""" ->
function initialise!(sampler::Sampler)
	for(i in 1:sampler.num_particles)
		sampler.particles[i] = Particle()
		from_prior!(sampler.particles[i])
		sampler.logl[i] = log_likelihood(sampler.particles[i])
		sampler.tiebreakers[i] = rand()
	end
	return nothing
end

@doc """
Find and save worst particle,
then generate replacement.
""" ->
function do_iteration!(sampler::Sampler, verbose::Bool)
	sampler.iteration += 1
	sampler.logx_threshold = -sampler.iteration/sampler.num_particles

	# Find index of worst particle
	worst = find_worst_particle(sampler::Sampler)

	## Get the particles with the lowest likelihood
	worst_particles = sampler.particles[worst]
	C = worst_particles.params[1]
	D = worst_particles.params[3]

	## Need to take the exponential of mu2 as it was modelled log(mu2)
	worst_particles.params[2] = exp(worst_particles.params[2])
	worst_particles.params[1] = worst_particles.params[1] * worst_particles.params[2] # mu1
	worst_particles.params[3] = worst_particles.params[3] * worst_particles.params[2] # L
	
	# Write its information to the output files
	if(sampler.iteration == 1)
		## Save temporary files "sample_info.txt" and "sample.txt"
		## (overwritten with each new player analysed)
		## Save named files "name_info.txt" and "name_"
		## sample_info.txt -
		f = open("sample_info.txt", "w")
		g = open(string(name, "_info.txt"), "w")
		write(f, "# num_particles, iteration, log(X), log(L)\n")
		write(g, "# num_particles, iteration, log(X), log(L)\n")
		## Save as files specified by 'name' in main.jl
		f2 = open("sample.txt", "w")
		g2 = open(string(name, "sample.txt"), "w")
		write(f2, "# The samples themselves. Use log(X) from sample_info.txt as un-normalised prior weights.\n")
		write(g2, "# The samples themselves. Use log(X) from sample_info.txt as un-normalised prior weights.\n")
	else
		f = open("sample_info.txt", "a")
		f2 = open("sample.txt", "a")
		g = open(string(name, "info.txt"), "a")
		g2 = open(string(name, "sample.txt"), "a")
	end
	write(f, string(sampler.num_particles), " ", string(sampler.iteration), " ",
			string(sampler.logx_threshold, " ", string(sampler.logl[worst]), "\n"))
	close(f)
	write(g, string(sampler.num_particles), " ", string(sampler.iteration), " ",
			string(sampler.logx_threshold, " ", string(sampler.logl[worst]), "\n"))
	close(g)
	write(f2, string(string(worst_particles), string(round(C, 5), " "), string(round(D, 5)), "\n")) # Write the worst_particles to sample.txt
	close(f2)
	write(g2, string(string(worst_particles), string(round(C, 5), " "), string(round(D, 5)), "\n"))
	close(g2)

	# Set likelihood threshold
	sampler.logl_threshold = sampler.logl[worst]
	if(verbose)
		println("Iteration ", sampler.iteration, ", log(X) = ",
				sampler.logx_threshold,	", log(L) = ", sampler.logl_threshold)
	end

	# Clone a survivor
	if(sampler.num_particles != 1)
		which = rand(1:sampler.num_particles)
		while(which == worst)
			which = rand(1:sampler.num_particles)
		end
		sampler.particles[worst] = deepcopy(sampler.particles[which])
		sampler.logl[worst] = deepcopy(sampler.logl[which])
		sampler.tiebreakers[worst] = deepcopy(sampler.tiebreakers[which])
	end

	# Evolve the particle
	accepted = 0::Int64
	for(i in 1:sampler.mcmc_steps)
		proposal = deepcopy(sampler.particles[worst])
		logH = perturb!(proposal)
		
		## This just makes sure if we have any values that are infinite, e.g. exp(100000) might give inf, have a numeric value. Exp(0) will give 1 - i.e. a definite acceptance
		if(logH > 0.0)
			logH = 0.0
		end

		logl_proposal = log_likelihood(proposal)
		tb_proposal = sampler.tiebreakers[worst] + randh()
		tb_proposal = mod(tb_proposal, 1.0)

		if((rand() <= exp(logH)) && is_less_than(
						(sampler.logl_threshold, sampler.tb_threshold),
						(logl_proposal, tb_proposal)))
			sampler.particles[worst] = proposal
			sampler.logl[worst] = logl_proposal
			sampler.tiebreakers[worst] = tb_proposal
			accepted += 1
		end
	end
	if(verbose)
		println("Accepted ", accepted, "/", sampler.mcmc_steps, " MCMC steps.\n")
	end

	return (sampler.logx_threshold, sampler.logl_threshold)
end

@doc """
Compare based on likelihoods first. Use tiebreakers to break a tie
""" ->
function is_less_than(x::Tuple{Float64, Float64}, y::Tuple{Float64, Float64})
	if(x[1] < y[1])
		return true
	end
	if((x[1] == y[1]) && (x[2] < y[2]))
		return true
	end
	return false
end


@doc """
Find the index of the worst particle.
""" ->
function find_worst_particle(sampler::Sampler)
	# Find worst particle
	worst = 1
	for(i in 2:sampler.num_particles)
		if(is_less_than((sampler.logl[i], sampler.tiebreakers[i]),
						(sampler.logl[worst], sampler.tiebreakers[worst])))
			worst = i
		end
	end
	return worst
end

@doc """
Calculate the log evidence, information, and posterior weights from the output of a run
""" ->
function calculate_logZ(logX::Vector{Float64}, logL::Vector{Float64})
	# Prior weights
	log_prior = logX - logsumexp(logX)
	# Unnormalised posterior weights
	log_post = log_prior + logL
	# log evidence and information
	logZ = logsumexp(log_post)
    post = exp(log_post - logZ)
	H = sum(post .* (log_post - logZ - log_prior))
	return (logZ, H, log_post)
end

@doc """
Do a Nested Sampling run.
""" ->
function do_nested_sampling(num_particles::Int64, mcmc_steps::Int64,
												depth::Float64; plot=true,
												verbose=true)

	# Number of NS iterations
	steps = Int64(max_depth*mcmc_steps)
	
	# Create the sampler
	sampler = Sampler(num_particles, mcmc_steps)
	initialise!(sampler)

	# Do 'steps' iterations of NS
	# Storage for results
	steps = Int64(max_depth)*num_particles
	plot_skip = num_particles

	# Store logX, logL
	keep = Array(Float64, (steps, 2))

	for(i in 1:steps)
		(keep[i, 1], keep[i, 2]) = do_iteration!(sampler, verbose)
		if(plot && (rem(i, plot_skip) == 0))
			(logZ, H, log_post) = calculate_logZ(keep[1:i, 1], keep[1:i, 2])
			uncertainty = sqrt(H/sampler.num_particles)

			PyPlot.subplot(2, 1, 1)
			PyPlot.hold(false)
			PyPlot.plot(keep[1:i, 1], keep[1:i, 2], "bo-", markersize=1)
			PyPlot.ylabel("\$\\ln(L)\$")
			PyPlot.title(string("\$\\ln(Z) =\$ ", signif(logZ, 6),
						" +- ", signif(uncertainty, 3),
						", \$H = \$", signif(H, 6), " nats"))

			# Adaptive ylim (exclude bottom 5%)
			logl_sorted = sort(keep[1:i, 2])
			lower = logl_sorted[1 + Int64(floor(0.05 * i))]
			PyPlot.ylim([lower, logl_sorted[end] + 0.05 * (logl_sorted[end] - lower)])

			PyPlot.subplot(2, 1, 2)
			PyPlot.plot(keep[1:i], exp(log_post - maximum(log_post)), "bo-", markersize = 1)
			PyPlot.xlabel("\$\\ln(X)\$")
			PyPlot.ylabel("Relative posterior weights")

			PyPlot.draw()
		end
	end

    results = calculate_logZ(keep[:,1], keep[:,2])
    post = exp(results[3] - results[1])
    writedlm("posterior_weights.txt", post)
	writedlm(string(name, "weights.txt"), post)

	if(verbose)
		println("Done!")
	end
	#if(PyPlot.plot)
		#PyPlot.ioff()
		#PyPlot.show()
		#end

end

File: sampler.cpp
// Nested sampling sampler

#include "sampler.h"
#include "model.h"
#include "Utils.h"

// Packages
#include <iostream>
#include <fstream> // for std::ofstream for writing to file

// Forward declare template
template class Sampler<Particle>;

// Define the Sampler Template class functions
// Sampler - constructor function
// Only takes a number of particles and mcmc steps
// Initiates empty vectors of particles and various log-likelihood thresholds
template <class T>
Sampler<T>::Sampler(std::vector<std::vector<int> > data, int nparticles, int mcmc_steps, std::string name) : 
	data(data), nparticles(nparticles), mcmc_steps(mcmc_steps), name(name),
	particles(std::vector<T> (nparticles)),
	logL(std::vector<double> (nparticles)),
	tiebreakers(std::vector<double> (nparticles))
{
	// Assign initial values
	iteration = 0;
	logX_threshold = 0.0;
	logL_threshold = 0.0;
	tb_threshold = 0.0;
}

// Sampler - initialise function
template <class T>
void Sampler<T>::initialise(RNG& rng)
{
	// Initialise the particles from the prior
	for (int i = 0; i < nparticles; i++)
	{
		particles[i].from_prior(rng); // Generate particle values from the prior
		logL[i] = particles[i].log_likelihood(data); // Calculate the log-likelihood values for each particle
		tiebreakers[i] = rng.rand();
	}
}

// Sampler - do_iteration - an iteration of nested sampling
template <class T>
std::vector<double> Sampler<T>::do_iteration(RNG& rng)
{	
	// Update current iteration and X threshold
	iteration += 1;
	double logX_threshold = static_cast<double>(-iteration)/nparticles;

	// Find index of worst particle in terms of likelihood
	int worst_index = find_worst_particle();

	// Get the worst particle and extract mu1, mu2 and L
	Particle worst = particles[worst_index];
	//double mu2 = exp(worst.params[1]);
	//double mu1 = worst.params[0] * mu2;
	//double L = worst.params[2] * mu2;

	//std::cout << "Worst Index = " << worst_index << ". Worst particle = " << particles[worst_index] << ", Log-likelihood = " << logL[worst_index] << std::endl;

	// Write the worst particle's information to file
	if(iteration == 1)
	{
		// Output file for samples
		std::ofstream outf_sample("sample.txt");
		std::ofstream save_sample(std::string("./Results/") + name + std::string("_sample.txt"));
		outf_sample << worst << std::endl;
		save_sample << worst << std::endl;
				
		// Output file for info
		std::ofstream outf_info("info.txt");
		std::ofstream save_info(std::string("./Results/") + name + std::string("_info.txt"));
		outf_info << nparticles << " " << iteration << "," << logX_threshold << "," << logL[worst_index] << std::endl;
		save_info << nparticles << " " << iteration << "," << logX_threshold << "," << logL[worst_index] << std::endl;
	}
	else
	{
		// Samples
		std::ofstream outf_sample;
		std::ofstream save_sample;
		outf_sample.open("sample.txt", std::ofstream::app);
		save_sample.open(std::string("./Results/") + name + std::string("_sample.txt"), std::ofstream::app);
		outf_sample << worst << std::endl;
		save_sample << worst << std::endl;

		// Info
		std::ofstream outf_info;
		std::ofstream save_info;
		outf_info.open("info.txt", std::ofstream::app);
		save_info.open(std::string("./Results/") + name + std::string("_info.txt"), std::ofstream::app);
		outf_info << nparticles << " " << iteration << "," << logX_threshold << "," << logL[worst_index] << std::endl;
		save_info << nparticles << " " << iteration << "," << logX_threshold << "," << logL[worst_index] << std::endl;
	}

	// Set the new log-likelihood threshold
	logL_threshold = logL[worst_index];
	std::cout << "Iteration " << iteration << ", Log(X) = " << logX_threshold << " , Log(L) = " << logL_threshold << std::endl;

	// Clone a surviving particle
	if(nparticles != 1)
	{
		// Randomly clone another particle
		int clone = rng.rand_int(nparticles);

		// Make sure we didn't clone the worst particle
		while(clone == worst_index)
		{
			clone = rng.rand_int(nparticles);
		}

		// Replace the worst particle with the worst particle
		particles[worst_index] = particles[clone];
		logL[worst_index] = logL[clone];
		tiebreakers[worst_index] = tiebreakers[clone];
	}

	// Evolve the particle for mcmc_steps so it is not identical to the particle it was cloned from
	int accepted = 0;
	for(int i = 0; i < mcmc_steps; i++)
	{
		// Propose a particle and perturb it
		Particle proposal = particles[worst_index];
		double log_h = proposal.perturb(rng);

		// Makes sure log_h isn't a silly value
		if(log_h > 0.0)
			log_h = 0.0;

		// Log-likelihood of proposed step
		double proposal_logL = proposal.log_likelihood(data);
		//std::cout << "LIKELIHOOD TEST: Particle = " << proposal << ", Likelihood = " << proposal_logL << std::endl;		
		double proposal_tb = fmod(tiebreakers[worst_index] + rng.randh(), 1.0);

		// Compare the proposal with the current worst threshold
		std::vector<double> current { proposal_logL, proposal_tb };
		std::vector<double> threshold { logL_threshold, tb_threshold };

		// Is the evolved particle's log-likelihood above the threshold?
		if((rng.rand() <= exp(log_h)) && is_less_than(threshold, current))
		{
			particles[worst_index] = proposal;
			logL[worst_index] = proposal_logL; 
			tiebreakers[worst_index] = proposal_tb;
			accepted += 1;
		}
	}

	// MCMC diagnostics
	std::cout << "Accepted " << accepted << " of " << mcmc_steps << " MCMC steps." << std::endl;
	//std::cout << "MCMC Particle = " << particles[worst_index] << ", Log-likelihood = " << logL[worst_index] << std::endl;
	//std::cout << std::endl;

	// Output to return
	std::vector<double> output { logX_threshold, logL_threshold };
	//std::cout << output[0] << ", " << output[1] << std::endl;

	return output;
}


// Find the index of the worst particle
template <class T>
int Sampler<T>::find_worst_particle()
{
	// Initialise the worst particle as the first
	int worst_index = 0;
	std::vector<double> worst{ logL[0], tiebreakers[0] };

	// Find the worst particle
	for(int i = 1; i < nparticles; i++)
	{
		// The current particle
		std::vector<double> current { logL[i], tiebreakers[i] };
		
		// Compare current and worst particle
		if(is_less_than(current, worst))
		{
			worst = current;
			worst_index = i;
		}
	}

	// Return the index of the worst particle
	return worst_index;
}



// Sampler - overload <<
// template <class T> - hmm this doesn't work for some reason, have to explcitly define << for Sampler<particle>
std::ostream& operator<<(std::ostream& out, const Sampler<Particle>& sampler)
{
	out << sampler.particles[0];
	return out;
}

// Sampler - print function (mostly so I can see what I've done is working)
template <class T>
void Sampler<T>::print()
{
	std::cout << "The sampler will use " << nparticles << " particles and " << mcmc_steps << " MCMC steps." << std::endl;
	std::cout << "Iteration #" << iteration << std::endl;
	std::cout << "Log-likelihood threshold: " << logL_threshold << std::endl;
	std::cout << "Log-likelihood X threshold: " << logX_threshold << std::endl;
	std::cout << "Log-likelihood tiebreaker threshold: " << tb_threshold << std::endl;
	std::cout << "Particle #1: " << particles[0] << std::endl;
	std::cout << "Log-likelihood #1: " << logL[0] << std::endl;
}



// Compare likelihoods of particles
// Pass two vectors of (likelihood, tiebreaker) as arguments for the current and worst particle
bool is_less_than(std::vector<double> current, std::vector<double> worst)
{
	// Current particle is worse than worst particle
	if(current[0] < worst[0])
		return true;
	// If likelihoods are the same, use tiebreakers
	if(current[0] == worst[0] && current[1] < worst[1])
		return true;
	
	// Otherwise worst particle is still the worst
	return false;
}


// Calculate the log evidence, information, and posterior weights from the output of a run
std::vector<std::vector<double> > calculate_logZ(const std::vector<double> logX, const std::vector<double> logL)
{
	// Storage vectors
	std::vector<double> log_prior(logX.size()); // prior weights
	std::vector<double> log_post(logX.size()); // unnormalised posterior weights
	double logZ; // log-evidence
	std::vector<double> post(logX.size()); // posterior weights
	std::vector<double> H_vec(logX.size()); // vector of information
	double H = 0.0; // information

	// Do logsumexp on the vector of logX
	double logX_store = logsumexp(logX);

	// Calculate for each iteration of the nested sampling run
	for(int i = 0; i < logX.size(); i++)
	{
		log_prior[i] = logX[i] - logX_store;
		log_post[i] = log_prior[i] + logL[i];
	}
	
	// Do logsumexp on the vector of log_post
	logZ = logsumexp(log_post);

	// Calcualte posterior weights
	for(int i = 0; i < logX.size(); i++)
	{
		post[i] = exp(log_post[i] - logZ);
		H_vec[i] = post[i] * (log_post[i] - logZ - log_prior[i]);
		H += H_vec[i]; // sum H_vec to get the information
	}

	// Return logZ, H and log_post
	std::vector<std::vector <double> > store(log_post.size(), std::vector<double> (3));

	// Populate store
	for(int i = 0; i < log_post.size(); i++)
	{
		store[i][0] = logZ;
		store[i][1] = H;
		store[i][2] = log_post[i];
	}
	
	return store;
}


// Do nested sampling function
void do_nested_sampling(const std::vector<std::vector<int> >& data, int nparticles, int mcmc_steps, double max_depth, std::string name, RNG& rng)
{
	// Number of nested sampling iterations
	int steps = nparticles * max_depth;
	
	// Create the Sampler and initialise it
	Sampler<Particle> sampler(data, nparticles, mcmc_steps, name);
	sampler.initialise(rng);
	
	// Storage of results
	std::vector<double> logX(steps);
	std::vector<double> logL(steps);

	// Do the nested sampling
	for(int i = 0; i < steps; i++)
	{
		std::vector<double> ns_run = sampler.do_iteration(rng);

		// Save log-X and log-likelihood thresholds
		logX[i] = ns_run[0];
		logL[i] = ns_run[1];
	}

	// Results and posterior weights
	std::vector<std::vector<double> > results = calculate_logZ(logX, logL);
	std::vector<double> post_weights(logX.size());

	// File to save the posterior weights
	std::ofstream outf("posterior_weights.txt");
	std::ofstream save_weights(std::string("./Results/") + name + std::string("_weights.txt"));

	// Save posterior weights to file
	for(int i = 0; i < results.size(); i++)
	{
		// Calculate weights
		post_weights[i] = exp(results[i][2] - results[i][0]);

		// Save weights to file
		std::ofstream save_weights;		
		save_weights.open(std::string("./Results/") + name + std::string("_weights.txt"), std::ofstream::app);
		
		outf << post_weights[i] << std::endl;
		save_weights << post_weights[i] << std::endl;
	}
}